lib/primitives/mtx.js

```diff
@@ -868,6 +868,53 @@ class MTX extends TX {
     return null;
   }
 
+  /**
+   * Sign custom P2WSH input.
+   * @param {Number} index - Index of input being signed.
+   * @param {KeyRing[]} rings - with private key.
+   * Sorted by desired signature order.
+   * @param {Script} redeemScript - its hash is in spending output.
+   * @param {Buffer|Number[]} witnessScript - signatures template.
+   * witnessScript is the SegWit equivalent of scriptSig.
+   * Put an empty buffer as placeholder for each signature.
+   * @param {SighashType} type
+   * @param {Coin|Output|Number} coin or output or value.
+   * @returns {Boolean} Whether the input was able to be signed.
+   */
+
+  signCustomInput(index, rings, redeemScript, witnessScript, type, coin) {
+    function getValue(tx, coin, input) {
+      if (coin) {
+        if (Coin.isCoin(coin) || Output.isOutput(coin))
+          return coin.value;
+
+        return coin;
```

Comment:
What is the expected type of `coin` in this branch? I'm guessing it would be a `Number` since the function is called `getValue` and `coin` is not a `Coin` and is not an `Output`.

Something like this might be able to simplify:

```js
if (!coin)
  coin = tx.view.getOutputFor(input);

assert(coin);
const value = coin.value;
```

Reply:
I needed to do the following instead, because otherwise the user can't give the value directly, but he'd have to give a coin or output.
```
    function getValue(tx, coin, input) {
      if (!coin)
        coin = tx.view.getOutputFor(input);
      assert(coin, 'Cannot get value of spending output');

      if (Number.isInteger(coin))
        return coin;

      return coin.value;
    }  
```

Reply:
Are you using this code in an environment where the coin object will not be available? `mtx.sign` assumes that the coinview is populated with coin objects. With access to a node, it should be possible to get the coin.

Reply:
Currently `mtx.signCustomInput()` has to be used directly by the user, as it's not called by `mtx.sign()` -- @pinheadmz [suggested something](https://github.com/bcoin-org/bcoin/pull/859?email_source=notifications&email_token=AEN5V6Z3US4D2KFACDY7NEDQLS7NLA5CNFSM4IZ42OP2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOD7VWS3Y#issuecomment-535521647) along these lines, but it would probably have to change the API of `mtx.sign()`.

Therefore it would be useful to allow the user maximum flexibility as to how to define the value of the spending output (hence the polymorphism). In particular, if the `input` is completely custom-built, it does not necessarily have the `prevout` property defined from which to extract the coin. The only information `mtx.signature()` needs from the coin is its value, thus if the user knows this value, why not let her pass it directly as a number?

(This also addresses your other comment)